[Blazor] Custom JS initializers

javiercn opened the pull request:
Adds support for custom JS initializers which allows to customize the Blazor boot process from JavaScript.

A concrete scenario is modifying how Blazor webassembly loads resources.
During the publish process, Blazor assets can be transformed and additional elements can be added to the manifest under the `extensions` path under resources. Like shown below:
```json
{
  "cacheBootResources": true,
  "config": [ ],
  "debugBuild": false,
  "entryAssembly": "blazorwasm-minimal",
  "icuDataMode": 0,
  "linkerEnabled": true,
  "resources": {
    "assembly": {
      "blazorwasm-minimal.dll": "sha256-psGoiPY+YcCvViu3IBI764Mj0zVP7wHYnMLoCdOn1xc=",
      ...
      "System.Threading.ThreadPool.dll": "sha256-xIFRbEifE\/hoQKqqxmQvQX\/FYSVnLf6Mi61CbK5GMVc="
    },
    "extensions": {
      "my-custom-extension": {
        "publish.extension.txt": "sha256-Ztb8hc\/OITLqsL7xtOlFeM+pfPIQ6A6bbbkJF2xHc0Q="
      }
    },
    "lazyAssembly": null,
    "libraryInitializers": {
      "blazorwasm-minimal.lib.module.js": "sha256-W5ZSepq1mBUmWnu7MqTAzO3YGJk7+LTHPaTOHFlY9Fw="
    },
    "pdb": null,
    "runtime": {
      "dotnet.timezones.blat": "sha256-Klww78mQczU8FGEQN5x4eSbjX8OwFHz7X\/OXmVKF8P8=",
      "dotnet.wasm": "sha256-VqE2ngWYg6zFZdUXTRkkNGpwPuryxZo3pRygj9iL\/Ko=",
      "icudt_CJK.dat": "sha256-8+0SUStQbNeoI3hDhcpyKdCWhsndM6zxj6jliMsXeIg=",
      "icudt_EFIGS.dat": "sha256-WqIury5JoQwiCF95WfagtEGBRK2mCvBDAzZEsinsL74=",
      "icudt_no_CJK.dat": "sha256-NepbHl0cZW1DE7TQbLOqzgNcIWnfgUq5RKaQL5ijw\/s=",
      "icudt.dat": "sha256-aAYuBDK\/9aSmctFbfPh62gWNg4mgsbu0rMqA08corAA=",
      "dotnet.6.0.0-rc.1.21376.24.js": "sha256-GSlngmlOn0UU8LJLn+t\/JNTul4dHENeq2y9M42lZNGU="
    },
    "satelliteResources": null
  }
}
```
The Blazor manifest includes the list of library initializers computed at build/publish time. Library initializers can export a couple of functions:
* `beforeBlazorStarts` which Blazor calls providing the options and the manifest and that can be use to tweak the options, for example, to replace the default bootResourceLoader.
* `afterBlazorStarted` which is called to let libraries know that Blazor has successfully started in case they need to do some initialization work afterwards.

Here is an example of the MSBuild required to "transform" the output. In this case, we are just emitting the output to a file and not transforming it at all, however you can replace this with whatever transform you want to perform. The target that we define does three things:
* Collects all Blazor static web assets
* Defines a new static web asset with the extension file
* Writes all the blazor assets to the "extension" file.

These new extension assets will automatically get included in the blazor.boot.json, compressed, and added to the service-worker-assets.json in case we are building a PWA.

They also contain the same hash data that is available to other assets, so caching and integrity can be maintained/implemented in the same way.

Finally, we could choose to remove all the dll entries on the manifest and just leave the extensions, however I believe its preferable to just extend the build output with more files and leave the existing ones there. The original dlls won't be downloaded by default (if you are not using a service worker) and if you are building a PWA, you can filter them out on the service-worker.js file.


**Update:**
I've simplified the process for Blazor webassembly as follows:
* There's an item group `PublishBlazorBootStaticWebAsset where all blazor resources have been collected.
* There's a BlazorPublishExtension itemgroup where you can define the files you want to add, specify the relative path and the extension name and we'll do the rest.

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly">
   ...
  <Target Name="_CustomizeBlazorBootProcess" AfterTargets="ProcessPublishFilesForBlazor">
    <ItemGroup>
      <!-- Define a new static web asset to make sure the new extension file is included in the output -->
      <BlazorPublishExtension Include="$(IntermediateOutputPath)publish.extension.txt">
        <RelativePath>_framework/publish.extension.txt</RelativePath>
        <ExtensionName>extension:my-custom-extension</ExtensionName>
      </BlazorPublishExtension>
      <FileWrites Include="$(IntermediateOutputPath)publish.extension.txt" />
    </ItemGroup>
    <WriteLinesToFile Lines="@(PublishBlazorBootStaticWebAsset)" File="$(IntermediateOutputPath)publish.extension.txt" WriteOnlyWhenDifferent="true" />
  </Target>
</Project>
```

javiercn committed:
Custom JS initializer sample

SteveSandersonMS commented on src/Components/Web.JS/tsconfig.json:
```diff
@@ -6,6 +6,7 @@
     "removeComments": false,
     "sourceMap": true,
     "target": "es2019",
+    "module": "es2020",
```
Out of interest, what's changed to require this?

javiercn replied:
I was getting some complaints with the dynamic import, and doesn't hurt to update it.

SteveSandersonMS replied:
Could you say which specific API feature it was complaining about? Would be good to check we're not losing compat with something like a recent version of iOS. I know we *say* we support "current" browsers but we still need to reason about times that we explicitly drop support for a major class of browsers.

javiercn replied:
TS1323: Dynamic imports are only supported when the '--module' flag is set to 'es2020', 'esnext', 'commonjs', 'amd', 'system', or 'umd'.

SteveSandersonMS replied:
Thanks, that explains it. I didn't understand why we had new requirements given that we already called `import` dynamically for the existing JS module loading feature. But I now see that `Microsoft.JSInterop`'s `tsconfig.json` includes:

```json
"module": "ESNext",
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/WebAssemblyStartOptions.ts:
```diff
@@ -10,11 +12,23 @@ export interface WebAssemblyStartOptions {
+   * The timeout in seconds before Blazor continues loading after invoking beforeBlazorStarts on a
+   * library initializer.
+   */
+  libraryInitializersTimeout?: number;
```
This seems like a sketchy concept. End users wouldn't be willing to wait for 30 seconds (the default timeout).

Do you think it would be OK to eliminate the timeout concept and just let developers be responsible for having their initializers complete in a sensible amount of time?

javiercn replied:
I think this is important to give developers final say on `how long` to wait until blazor continues loading, this is to `discourage` library authors from writing initializers that slowdown the startup experience.

Essentially, a library can't slowdown blazor loading nor prevent it from starting after X seconds. If you think we don't need this, I'm happy to take this out.

SteveSandersonMS replied:
While it might be OK to have a timeout that throws and kills the whole startup process if an initializer is taking too long, I don't think it would normally be OK to just ignore the initializer because it's taking too long. The initializer must be there for a reason, so the application would be in an invalid state if it hasn't yet completed.

I would rather just not have a timeout, so if an initializer is behaving badly, that's a problem that the initializer author and the application developer (who may be the same person) have to sort out between them. We shouldn't take an opinion that it's safe to skip waiting for the initializer.

javiercn replied:
I've simplified the logic here to just call on the functions and await them all without a timeout

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/WebAssemblyStartOptions.ts:
```diff
@@ -10,11 +12,23 @@ export interface WebAssemblyStartOptions {
+   * Overrides the built-in way of loading blazor.boot.json
+   * @param environment The built-in environment setting on start.
+   */
+  loadBlazorBootJson(environment?: string) : BootConfigResult;
```
What do you think about possibly merging this into the existing `loadBootResource` callback? We could add an extra `environment` parameter to it.

The existing semantics for `loadBootResource` are pretty good about falling back on default behavior for `type` values that are not known to the callback - they just return null/undefined and we fall back on the default.

javiercn replied:
I've integrated this with the original loadBootResource callback.

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.ts:
```diff
@@ -116,7 +119,54 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
+          },
+            timeout))]);
+    }
+  }
```
Can we move all this new logic out into a separate file? The Boot.*.ts files are already a big collection of semi-related stuff so it would be great to minimize the amount of new logic that shows up all in one place.

Since an equivalent is needed in Server/WebView too, this would also help to avoid duplicating definitions for all the types.

javiercn replied:
done

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.ts:
```diff
@@ -116,7 +119,54 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
     }
   };
 
+  type BeforeBlazorStartedCallback = (options: Partial<WebAssemblyStartOptions> | undefined, bootConfigResul: BootConfigResult) => Promise<void>;
```
```suggestion
  type BeforeBlazorStartedCallback = (options: Partial<WebAssemblyStartOptions> | undefined, bootConfigResult: BootConfigResult) => Promise<void>;
```

javiercn replied:
I've cleaned these up to make them `platform` independent. The caller flows the right set of arguments to them.

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.ts:
```diff
@@ -116,7 +119,54 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
 
+  type BeforeBlazorStartedCallback = (options: Partial<WebAssemblyStartOptions> | undefined, bootConfigResul: BootConfigResult) => Promise<void>;
+  type AfterBlazorStartedCallback = () => Promise<void>;
+  type BlazorInitializer = { beforeBlazorStarts: BeforeBlazorStartedCallback, afterBlazorStarted: AfterBlazorStartedCallback };
```
In terms of naming, do you think it would be reasonable to change to something like the following?

```suggestion
  type BlazorInitializer = { beforeStart: BeforeBlazorStartedCallback, afterStart: AfterBlazorStartedCallback };
```

javiercn replied:
I've done this change, however I'm a bit worried about using such common names. If you don't share that concern, let's keep it this way.

javiercn replied:
(tp be clear, I mean with your changes)

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.ts:
```diff
@@ -128,7 +178,12 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
+  async function invokeBlazorStartedCallbacks(applicationStarted: Promise<void>) {
+    await applicationStarted;
+    await Promise.all(afterBlazorStartedCallbacks.map(c => c()));
+  }
```
Not sure why this is a function - can it be inlined into the code above?

SteveSandersonMS replied:
Oh is it because you don't want the returned promise to include these awaits?

javiercn replied:
Yep

SteveSandersonMS commented:
One concern I have here is that this makes the precise format of `blazor.boot.json` become part of the public API surface. That was never the case previously, and means we'll potentially face significant limitations in how we can evolve it in the future.

To mitigate this, would it be reasonable to change the info we pass to the callbacks to *only* include the `extensions` part of the data? This way even if we need to completely change the notion of `blazor.boot.json` in the future we can still do so as long as there remains some way to get the `extensions` data.

This might also make it a lot more practical to set up a package that includes an initializer that works on WebAssembly and Server, where Server could also have this `extensions` notion in some way if we wanted but wouldn't have the rest of `blazor.boot.json`. But even if this is not the case, it's still valuable to to make `blazor.boot.json` become public API.

SteveSandersonMS commented:
Regarding your example MSBuild code above, could you clarify which parts of it refer to public, documented, supported APIs? For example:

 * The `AssetTraitName` and the specific values we see for `AssetTraitValue` (`BlazorWebAssemblyResource`, `Culture`, etc.) - would a developer be able to find docs listing the possibilities there, and would it count as breaking if we added further ones in the future, since then the logic here that filters out based on `AssetTraitValue` would start including extra things?
 * This string format: `extension:my-custom-extension`
 * What happens if you want to have multiple "asset traits"? Is that meaningful?

> The original dlls won't be downloaded by default

What stops the original DLLs being downloaded by default?

javiercn committed:
Option 1

javiercn committed:
tmp

javiercn committed:
undo changes on loader

javiercn committed:
Addresses feedback from Steve

javiercn committed:
Cleanup

javiercn committed:
Cleanups and Blazor server implementation

javiercn commented:
> One concern I have here is that this makes the precise format of `blazor.boot.json` become part of the public API surface. That was never the case previously, and means we'll potentially face significant limitations in how we can evolve it in the future.
> 
> To mitigate this, would it be reasonable to change the info we pass to the callbacks to _only_ include the `extensions` part of the data? This way even if we need to completely change the notion of `blazor.boot.json` in the future we can still do so as long as there remains some way to get the `extensions` data.
> 
> This might also make it a lot more practical to set up a package that includes an initializer that works on WebAssembly and Server, where Server could also have this `extensions` notion in some way if we wanted but wouldn't have the rest of `blazor.boot.json`. But even if this is not the case, it's still valuable to to make `blazor.boot.json` become public API.

I've modified the process to just pass the `extensions` part.

javiercn commented:
> * The `AssetTraitName` and the specific values we see for `AssetTraitValue` (`BlazorWebAssemblyResource`, `Culture`, etc.) - would a developer be able to find docs listing the possibilities there, and would it count as breaking if we added further ones in the future, since then the logic here that filters out based on `AssetTraitValue` would start including extra things?
> * This string format: `extension:my-custom-extension`
> * What happens if you want to have multiple "asset traits"? Is that meaningful?

I'm going to give people an ItemGroup with these so that they don't have to worry about them.

javiercn commented:
> What stops the original DLLs being downloaded by default?

Your custom loader does by downloading whatever package format you have ahead of time (during the initializer phase), parsing it and creating an already made list of responses that matches that of the requested dlls

javiercn committed:
Webassembly fixes

javiercn committed:
Add blazor server support

javiercn committed:
Partial cleanup of the static web assets implementation

javiercn committed:
Support for Blazor webview

javiercn committed:
Fix tests

javiercn committed:
Add E2E tests

javiercn committed:
Sample JS module sample

javiercn commented:
Ping

SteveSandersonMS commented on src/Components/Samples/BlazorServerApp/wwwroot/BlazorServerApp.lib.module.js:
```diff
@@ -0,0 +1,7 @@
+
+export function afterStarted(blazor) {
+    console.log('started!');
+}
```
Would it be OK to not add this file? Generally the "playground" apps (of which this is one) get really hard to use if they become filled up with examples of random features. Where possible it's great to keep this project and the WebAssembly standalone project close to the project template.

javiercn replied:
absolutely, just did it to make sure I could test things (since I had to monkeypatch my sdk to do this) and there are also E2E test for this.

SteveSandersonMS commented on src/Components/Server/src/PublicAPI.Unshipped.txt:
```diff
@@ -1,4 +1,5 @@
 #nullable enable
+Microsoft.AspNetCore.Components.Server.CircuitOptions.JavaScriptInitializers.get -> System.Collections.Generic.IList<string!>!
```
How come this is part of the public API? I thought the list of initializers was determined by the presence of `.js` files on disk. Under what circumstances would a developer mutate the list at runtime by setting things in these options?

javiercn replied:
I can make this an implementation detail, I think it might be ok if we let folks tweak the list of things that we will autoload.

SteveSandersonMS replied:
I don't actually mind which way you go with this, but it still appears to be an open question.

javiercn replied:
Let's discuss this in this afternoon's standup and make a decision on it.

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebView.ts:
```diff
@@ -13,6 +14,16 @@ async function boot(): Promise<void> {
+
+  const initializers: string[] = await jsInitializersResponse.json();
+  const jsInitializer = new JSInitializer();
+  await jsInitializer.importInitializersAsync(initializers, []);
```
Any chance of moving the repeated server/webview code (and possible more of the webassembly code) out into a separate place?

Apologies if this is a bit fussy, but these initialization files just keep on growing with every new feature and I feel like I have to try to hold back the tide on it. The WebView one was pretty small and clean and would be great to keep like that.

SteveSandersonMS replied:
Hopefully would only take a few mins, as you already have the `JSInitializers.ts` file and can export another function from it.

javiercn replied:
I thought I did this, do you have something even more 'minimalistic' in mind?

It's 4 lines total compared to 2 lines (which is the min I can reduce it to)

SteveSandersonMS commented on src/Components/Web.JS/src/GlobalExports.ts:
```diff
@@ -13,7 +13,7 @@ import { getNextChunk, receiveDotNetDataStream } from './StreamingInterop';
 import { attachWebRendererInterop } from './Rendering/WebRendererInteropMethods';
 
-interface IBlazor {
+export interface IBlazor {
```
I think you could use the `Blazor` const in your other function rather than `IBlazor`. TBH `IBlazor` doesn't serve any purpose and we should ideally eliminate it some time in the future.

javiercn replied:
I'll give it a try

javiercn replied:
This doesn't really work, since I needed it at a type position within the other file. I've chosen to type the functions as `unknown` since the only thing we do is pass through data

javiercn replied:
Nevermind, I found the way

SteveSandersonMS commented on src/Components/Web.JS/src/JSInitializers.ts:
```diff
@@ -0,0 +1,53 @@
+    try {
+      await Promise.all(initializerFiles.map(f => importAndInvokeInitializer(this, f)));
+    } catch (error) {
+      console.warn(`A library initializer produced an error before starting: '${error}'`);
```
What's the reason for swallowing the exception? Shouldn't we let initialization fail?

SteveSandersonMS replied:
Same with the other try/catch below.

javiercn replied:
Fair enough, I guess folks need to be responsible for what they bring into their apps, and deal with it if it crashes their app.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/BootConfig.ts:
```diff
@@ -1,13 +1,17 @@
+import { WebAssemblyBootResourceType } from "./WebAssemblyStartOptions";
```
Nitpick, please ignore if you want. I know the tooling will have done this for you.

```suggestion
import { WebAssemblyBootResourceType } from './WebAssemblyStartOptions';
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/BootConfig.ts:
```diff
@@ -1,13 +1,17 @@
-      cache: 'no-cache'
-    });
+  static async initAsync(loadBootResource?: (type: WebAssemblyBootResourceType, name: string, defaultUri: string, integrity: string) => string | Promise<Response> | null | undefined, environment?: string): Promise<BootConfigResult> {
+    let loaderResponse = loadBootResource !== undefined ?
+      loadBootResource('manifest', 'blazor.boot.json', '_framework/blazor.boot.json', '') :
+      this.defaultLoadBlazorBootJson('_framework/blazor.boot.json');
```
Another nit, to match code style in other TS places:

```suggestion
    let loaderResponse = loadBootResource !== undefined
      ? loadBootResource('manifest', 'blazor.boot.json', '_framework/blazor.boot.json', '')
      : this.defaultLoadBlazorBootJson('_framework/blazor.boot.json');
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/BootConfig.ts:
```diff
@@ -1,13 +1,17 @@
-      credentials: 'include',
-      cache: 'no-cache'
-    });
+  static async initAsync(loadBootResource?: (type: WebAssemblyBootResourceType, name: string, defaultUri: string, integrity: string) => string | Promise<Response> | null | undefined, environment?: string): Promise<BootConfigResult> {
```
This is a hyper-complicated type definition. Could you declare a type or interface for some of this to make it short enough to be readable?

For example for the `(type: WebAssemblyBootResourceType, name: string, defaultUri: string, integrity: string) => string | Promise<Response> | null | undefined` part

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/WebAssemblyStartOptions.ts:
```diff
@@ -1,3 +1,5 @@
+import { BootConfigResult, BootJsonData } from "./BootConfig";
```
I wonder if there's a .editorconfig for this

```suggestion
import { BootConfigResult, BootJsonData } from './BootConfig';
```

javiercn replied:
We had TSLint enabled at some point, but I think we end up ignoring it.

SteveSandersonMS commented on src/Components/WebAssembly/Samples/HostedBlazorWebassemblyApp/Client/wwwroot/HostedBlazorWebassemblyApp.Client.lib.module.js:
```diff
@@ -0,0 +1,7 @@
+
+export function afterStarted(blazor) {
+    console.log('started!');
+}
```
Could this be removed from the sample? Should be enough to have it in the E2E test app.

SteveSandersonMS commented on src/Components/WebView/Samples/PhotinoPlatform/src/BlazorWindow.cs:
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Reflection;
```
Unneeded?

javiercn replied:
I hate auto-namespace import for this reason.

SteveSandersonMS commented on src/Components/WebView/Samples/PhotinoPlatform/testassets/PhotinoTestApp/wwwroot/PhotinoTestApp.lib.module.js:
```diff
@@ -0,0 +1,7 @@
+
+export function afterStarted(blazor) {
+    console.log('started!');
+}
```
Doesn't look like this is used in the E2E test, so could it be removed?

SteveSandersonMS commented on src/Components/WebView/WebView/src/blazor.modules.json:
```diff
@@ -0,0 +1 @@
+[]
```
I'm curious about what's going on here. Could you explain?

javiercn replied:
There is no manifest file when there are no modules, so we embed an empty one that gets served as a fallback (same way as how the `_framework/blazor.webview.js`). The alternative (if we want to avoid 404s on the browser) is to special case this path in the StaticContentProvider (which I wanted to avoid because it seemed more hassle).

javiercn replied:
This is in essence the Webview equivalent of the "blazor server endpoint"

SteveSandersonMS commented on src/Components/test/E2ETest/ServerExecutionTests/JsInitializersTest.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests
+{
+    public class ServerJsInitializersTest : ServerTestBase<ToggleExecutionModeServerFixture<Program>>
```
This looks the same as `JsInitializersTest`. Does it work to subclass it for the server, like we do for most of the E2E tests?

javiercn replied:
Ah, copy paste stuff, I'll fix it.

SteveSandersonMS commented on src/Components/test/testassets/BasicTestApp/wwwroot/BasicTestApp.lib.module.js:
```diff
@@ -0,0 +1,48 @@
+</tr>
+${rows}
+`;
+            document.body.appendChild(requestTable);
```
This is cool!

SteveSandersonMS commented on src/Hosting/Hosting/src/StaticWebAssets/StaticWebAssetsLoader.cs:
```diff
@@ -93,12 +65,11 @@ internal static (Stream, bool) ResolveManifest(IWebHostEnvironment environment,
             var basePath = string.IsNullOrEmpty(assembly.Location) ? AppContext.BaseDirectory : Path.GetDirectoryName(assembly.Location);
-            yield return (Path.Combine(basePath!, $"{environment.ApplicationName}.staticwebassets.runtime.json"), isJson: true);
-            yield return (Path.Combine(basePath!, $"{environment.ApplicationName}.StaticWebAssets.xml"), isJson: false);
+            return Path.Combine(basePath!, $"{environment.ApplicationName}.staticwebassets.runtime.json");
```
I don't have context on the changes in this file so I'll just assume it's all great 👍

javiercn replied:
Yeah, this is just cleanup from preview7. Since Blazor Desktop and ASP.NET Core consume the same manifest now I did the cleanup in one go to avoid having to update stuff that I would have to delete.

javiercn commented on src/Components/Server/src/Circuits/CircuitOptionsJavaScriptInitializersConfiguration.cs:
```diff
@@ -0,0 +1,33 @@
+                for (var i = 0; i < initializers.Length; i++)
+                {
+                    var initializer = initializers[i];
+                    options.JavaScriptInitializers.Add(initializer);
```
For 2 reaons:
* It doesn't get generated when there are no modules, and want to avoid the 404.
* The manifest includes the app name, which would need to be passed down to JS otherwise. We don't know if the manifest is for a blazor server app or if an MVC app wants to consume it (that's our contract for MVC/RP) so the endpoint abstracts over that.

I tried to do it through the websocket connection, but the ordering in which things happen makes it problematic.

SteveSandersonMS replied:
The 404 reason makes sense. Thanks.

Presumably people could actually fetch `AppName.manifest.json` (or whatever it's called) if they knew the URL to it within `wwwroot`, but we don't really care before it's only the same info they can get through the middleware anyway.

javiercn committed:
Address feedback

javiercn committed:
Enable dynamic import

javiercn committed:
Fix build

javiercn committed:
Fix bad merge

javiercn committed:
Address missing feedback

javiercn commented:
🆙 📅

javiercn committed:
Fix build break

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.Server.ts:
```diff
@@ -77,6 +79,8 @@ async function boot(userOptions?: Partial<CircuitStartOptions>): Promise<void> {
 
   logger.log(LogLevel.Information, 'Blazor server-side application started.');
+
+  jsInitializer.invokeAfterStartedCallbacks(Blazor);
```
This is nice. Thanks for keeping it minimal here :)

SteveSandersonMS commented on src/Components/Web.JS/src/JSInitializers/JSInitializers.ts:
```diff
@@ -0,0 +1,43 @@
+import { Blazor } from '../GlobalExports'
+type IBlazor = typeof Blazor;
```
This is unusual - having the `I` prefix is a C#-ism that we don't normally see in TypeScript, and it isn't an interface here either.

If you're making any further updates in this PR it would be nice to just inline the `typeof Blazor` expression into the two usages below, but I'm very aware this is just nitpicky and doesn't actually matter, so please feel free to ignore it if you're not making further updates :)

SteveSandersonMS replied:
Also while I'm doing nitpicks, there's a missing semicolon at the end of line 1.

SteveSandersonMS commented on src/Components/Web.JS/src/JSInitializers/JSInitializers.ts:
```diff
@@ -0,0 +1,43 @@
+  async importInitializersAsync(
+    initializerFiles: string[],
+    initializerArguments: unknown[]): Promise<void> {
+      await Promise.all(initializerFiles.map(f => importAndInvokeInitializer(this, f)));
```
Indentation looks off here.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/BootConfig.ts:
```diff
@@ -1,13 +1,20 @@
+
+    const bootConfigResponse = loaderResponse instanceof Promise ?
+      await loaderResponse :
+      await BootConfigResult.defaultLoadBlazorBootJson(loaderResponse ?? '_framework/blazor.boot.json');
```
Is there any distinction between `this.defaultLoadBlazorBootJson` (on line 13) and `BootConfigResult.defaultLoadBlazorBootJson` (on line 17)? I think they are the same thing. If so, to avoid confusion, it would be ideal to reference the function in the same way both times. Apologies if I'm missing some reason why it needs to vary.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/BootConfig.ts:
```diff
@@ -17,8 +24,17 @@ export class BootConfigResult {
     return new BootConfigResult(bootConfig, applicationEnvironment);
   };
+
+  public static async defaultLoadBlazorBootJson(url: string) : Promise<Response> {
```
Looks like this could just be a regular `function` - doesn't need to be publicly reachable from `BootConfigResult` AFAICT.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/BootConfig.ts:
```diff
@@ -17,8 +24,17 @@ export class BootConfigResult {
+  }
 }
 
+
```
```suggestion
```

SteveSandersonMS commented on src/Components/test/testassets/TestContentPackage/ComponentFromPackage.razor:
```diff
@@ -1,11 +1,30 @@
+
+    async Task TriggerPrompt()
+    {
+        await _module.InvokeVoidAsync("showPrompt", "Hello from module");
```
This doesn't seem to be used by the E2E test code in `JsInitializersTest`, unless I'm missing something. Do you think it would be reasonable to include it in the E2E test coverage? Or if not, should it be removed?

javiercn replied:
Changed it and included a test to validate it

SteveSandersonMS reviewed (approved):
Excellent. Looking forwards to us shipping this!

I added a few comments, but they are minor details.

javiercn committed:
Address feedback

javiercn committed:
Make public API internal

javiercn committed:
Add additional E2E test for file modules

javiercn committed:
Temporarily supress a bunch of warnings

javiercn committed:
Add additional test

javiercn commented:
I've addressed the remaining feedback and I'm just waiting for the build to do it's thing at this point. 

The last bit is to figure out the linker issues that have surfaced with the latest SDK bits and which seem totally unrelated to these changes.

javiercn commented:
We need to update the runtime version too, I believe we are running into https://github.com/dotnet/roslyn/issues/55461

javiercn committed:
Update JS files

javiercn committed:
Update identity tests
